Use more smart pointers in CSS parsing code

cdumez opened the pull request:
#### f0a2432b920fc757138bcc928624389e773ba047
<pre>
Use more smart pointers in CSS parsing code
<a href="https://bugs.webkit.org/show_bug.cgi?id=261578">https://bugs.webkit.org/show_bug.cgi?id=261578</a>

Reviewed by Darin Adler and Brent Fulgham.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeFontPaletteValuesRule):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):
(WebCore::isValueIDPair):
(WebCore::isNumber):
(WebCore::isValueID):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::backgroundPositionFromThreeValues):
(WebCore::CSSPropertyParserHelpers::positionFromFourValues):
(WebCore::CSSPropertyParserHelpers::consumePositionCoordinates):
(WebCore::CSSPropertyParserHelpers::isGridTrackFixedSized):
* Source/WebCore/css/parser/SizesAttributeParser.cpp:
(WebCore::SizesAttributeParser::calculateLengthInPixels):
(WebCore::SizesAttributeParser::mediaConditionMatches):
(WebCore::SizesAttributeParser::parse):
(WebCore::SizesAttributeParser::effectiveSizeDefaultValue):
(WebCore::SizesAttributeParser::protectedDocument const):
* Source/WebCore/css/parser/SizesAttributeParser.h:

Canonical link: <a href="https://commits.webkit.org/268048@main">https://commits.webkit.org/268048@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7210686dd1daa95d74823664ca6262fb28302c99

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18804 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19399 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20315 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17282 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18660 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22110 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18953 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19176 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18689 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18872 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16082 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21195 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16095 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16840 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23315 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17115 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17010 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21208 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17592 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14921 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16661 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4401 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -818,20 +818,18 @@ RefPtr<StyleRuleFontPaletteValues> CSSParserImpl::consumeFontPaletteValuesRule(C
-        if (!cssPrimitiveFontFamily)
+        if (RefPtr cssPrimitiveFontFamily = dynamicDowncast<CSSPrimitiveValue>(*cssFontFamily)) {
+            append(*cssPrimitiveFontFamily);
             return fontFamilies;
```
It doesn't seem like this early return is needed with your new code path, since it should just fall through to returning `fontFamilies`.

brentfulgham commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1234,7 +1234,7 @@ static constexpr InitialValue initialValueForLonghand(CSSPropertyID longhand)
 static bool isValueIDPair(const CSSValue& value, CSSValueID valueID)
 {
-    return value.isPair() && isValueID(value.first(), valueID) && isValueID(value.second(), valueID);
+    return value.isPair() && isValueID(Ref { value.first() }.get(), valueID) && isValueID(Ref { value.second() }.get(), valueID);
```
According to `CSSPrimitiveValue.h`, the `isValueID(...)` overloads accept Ref and RefPtr directly, so no need for the `get()` calls.

But I also wonder if it's worth `Ref` here, as the method just unrolls to `<argument>.valueID()`

brentfulgham replied:
I.e.,:
`isValueID(value.first(), valudID)`

Just becomes:
`value.first().valueID() == valueID`

cdumez replied:
The clang analyzer complains about it, which is why I am doing it. Now we can choose to ignore the clang analyzer but I am not sure what our policy is in this regards.

I'll try without the `.get()` but if I remember correctly, I had to add them because the compiler complained the call was ambiguous.

cdumez replied:
```
In file included from WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource132.cpp:8:
Source/WebCore/css/parser/CSSPropertyParser.cpp:1237:30: error: call to 'isValueID' is ambiguous
    return value.isPair() && isValueID(Ref { value.first() }, valueID) && isValueID(Ref { value.second() }, valueID);
WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource132.cpp:1:
In file included from Source/WebCore/css/parser/CSSParserFastPaths.cpp:36:
Source/WebCore/css/CSSPrimitiveValue.h:366:13: note: candidate function
inline bool isValueID(const RefPtr<CSSPrimitiveValue>& value, CSSValueID id)
Source/WebCore/css/CSSPrimitiveValue.h:371:13: note: candidate function
inline bool isValueID(const Ref<CSSPrimitiveValue>& value, CSSValueID id)
Source/WebCore/css/CSSPrimitiveValue.h:376:13: note: candidate function
inline bool isValueID(const CSSValue& value, CSSValueID id)
Source/WebCore/css/CSSPrimitiveValue.h:386:13: note: candidate function
inline bool isValueID(const RefPtr<CSSValue>& value, CSSValueID id)
```

darinadler replied:
We should consider fixing that overloading so it's unambiguous. But also, if `isValueID` is causing us to have to ref things then we should get rid of it. Its purpose is to make the idiom of checking if a value is a specific keyword more convenient and it's super strange if that requires reference count churn.

brentfulgham commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3382,30 +3382,31 @@ static std::optional<PositionCoordinates> positionFromTwoValues(CSSPrimitiveValu
 
-    CSSPrimitiveValue* center = nullptr;
+    RefPtr<CSSPrimitiveValue> center;
     for (int i = 0; values[i]; i++) {
```
Perhaps this should be `++i` ?

brentfulgham reviewed (approved):
Looks good to me! r=me

darinadler commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -818,20 +818,16 @@ RefPtr<StyleRuleFontPaletteValues> CSSParserImpl::consumeFontPaletteValuesRule(C
-        if (!cssPrimitiveFontFamily)
-            return fontFamilies;
-        append(*cssPrimitiveFontFamily);
+        if (RefPtr cssPrimitiveFontFamily = dynamicDowncast<CSSPrimitiveValue>(*cssFontFamily))
```
`releaseNonNull` instead of `*` perhaps

Also I would name the local something shorter like `family`.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -494,9 +494,9 @@ RefPtr<CSSCustomPropertyValue> CSSPropertyParser::parseTypedCustomPropertyValue(
-        auto& valueList = downcast<CSSValueContainingVector>(*value);
-        auto syntaxValueList = CSSCustomPropertyValue::SyntaxValueList { { }, valueList.separator() };
-        for (auto& listValue : valueList) {
+        Ref valueList = downcast<CSSValueContainingVector>(*value);
```
`releaseNonNull` instead of `*`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1254,18 +1254,18 @@ static bool isNumber(const CSSValue& value, double number, CSSUnitType type)
-        && isNumber(quad.right(), number, type)
-        && isNumber(quad.bottom(), number, type)
-        && isNumber(quad.left(), number, type);
+    return isNumber(Ref { quad.top() }.get(), number, type)
```
This is not good. The point of these functions like `isNumber` is to create a simple caller interface so functions like this are brief and easy to read. If callers have to keep doing `Ref` we should change the idiom. The idea is that these inline functions are "free" and easy to use and result in readable code.

I understand that we are just doing what the static analyzer says but it's going against why we wrote the code the way we did.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7769,8 +7771,8 @@ static bool isGridTrackFixedSized(const CSSValue& value)
         return false;
-    return isGridTrackFixedSized(downcast<CSSPrimitiveValue>(*function.item(0)))
-        || isGridTrackFixedSized(downcast<CSSPrimitiveValue>(*function.item(1)));
+    return isGridTrackFixedSized(Ref { downcast<CSSPrimitiveValue>(*function.item(0)) })
```
Same as above. If callers have to do `Ref` then maybe we should just change the argument type.

cdumez committed:
Use more smart pointers in CSS parsing code
https://bugs.webkit.org/show_bug.cgi?id=261578

Reviewed by Darin Adler and Brent Fulgham.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeFontPaletteValuesRule):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseTypedCustomPropertyValue):
(WebCore::isValueIDPair):
(WebCore::isNumber):
(WebCore::isValueID):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::backgroundPositionFromThreeValues):
(WebCore::CSSPropertyParserHelpers::positionFromFourValues):
(WebCore::CSSPropertyParserHelpers::consumePositionCoordinates):
(WebCore::CSSPropertyParserHelpers::isGridTrackFixedSized):
* Source/WebCore/css/parser/SizesAttributeParser.cpp:
(WebCore::SizesAttributeParser::calculateLengthInPixels):
(WebCore::SizesAttributeParser::mediaConditionMatches):
(WebCore::SizesAttributeParser::parse):
(WebCore::SizesAttributeParser::effectiveSizeDefaultValue):
(WebCore::SizesAttributeParser::protectedDocument const):
* Source/WebCore/css/parser/SizesAttributeParser.h:

Canonical link: https://commits.webkit.org/268048@main
